### PR TITLE
ws/pgwire: Notify of changes to Session Vars

### PR DESCRIPTION
**Pull request by ParkMyCar**
### Motivation

This PR does a few things:

1. Updates the "notify_set" to include "cluster" and "cluster_replica".
2. Updates the web socket protocol to include `ParameterStatus` messages, so we can notify clients when a session parameter changes.
3. Update the pgwire protocol to send `ParameterStatus` messages when a transaction ends, if any session variables changed with the end of the transaction.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - This PR updates the web socket protocol so clients can get notified when session parameters change.

### CODE REVIEW

**Comment by mjibson** [comment]
You should be able to use a similar technique from https://github.com/MaterializeInc/materialize/blob/8e94b9b5be22b90d16c9abe456d7a64390ecd74e/src/pgwire/src/protocol.rs#L1328 instead of adding a channel. Can add that at https://github.com/MaterializeInc/materialize/blob/8e94b9b5be22b90d16c9abe456d7a64390ecd74e/src/pgwire/src/protocol.rs#L1328

**Comment by benesch** [comment]
Came by to drop the same thought! The trick of just sniffing out responses to explicit `SET` commands has served us well in the pgwire layer.

That said, in the fullness of time, I imagine we'll need something like this! `ALTER SYSTEM` can change session state for all sessions, and I can imagine wanting to flow those updates all the way through to the clients.

**Comment by ParkMyCar** [comment]
@mjibson and @benesch thanks for the early review! I updated the approach to use `ExecuteResponse` like you mentioned, let me know what you think and if it looks good I'll come back and add more tests

**Comment by benesch** [comment]
This makes sense to me, but deferring actual review to Matt!

File: src/environmentd/src/http/sql.rs [hunk 1]
```diff
@@ -863,17 +882,32 @@ async fn execute_stmt<S: ResultSender>(
...
+            // When a transaction ends there is a chance our variables change, so send everything
+            // in the notify set.
+            let params = client
+                .session()
+                .vars()
+                .notify_set()
+                .map(|v| ParameterStatus { name: v.name().to_string(), value: v.value() })
+                .collect();
+            SqlResult::ok(client, tag.expect("ok only called on tag-generating results"), params).into()
+        },
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/environmentd/src/http/sql.rs:903]: Can you make an issue or PR to do that too?

**Review by mjibson — APPROVED** [review]
LGTM.

But, we need some followup work to handle these:

```
BEGIN;
SET z = 'a';
ROLLBACK;
```

```
BEGIN;
SET LOCAL z = 'a';
COMMIT;
```

In both of these cases, the variable change message is sent, but after the txn ends the variable reverts to something else, and that change isn't sent.

File: src/environmentd/tests/server.rs [hunk 2]
```diff
@@ -1420,3 +1414,37 @@ fn test_concurrent_id_reuse() {
...
+        serde_json::from_str(&msg).unwrap()
+    };
+
+    let complete = read_msg();
+    let read = read_msg();
+    let param_update = read_msg();
+
+    assert!(matches!(complete, WebSocketResponse::CommandComplete(_)));
+    assert!(matches!(read, WebSocketResponse::ReadyForQuery(_)));
+    assert!(matches!(param_update, WebSocketResponse::ParameterStatus(_)));
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/environmentd/tests/server.rs]: I moved these cases into the `ws` test file, which asserts their values

File: src/environmentd/src/http/sql.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/environmentd/src/http/sql.rs]: I ended up just including this change in this PR since it was simple enough

**Comment by ParkMyCar** [comment]
@mjibson would you mind taking another look? I made a decent amount of changes. This now includes the changes for pgwire, and when a transaction ends only sending `ParameterStatus` messages for parameters that actually changed, instead of sending the entire notify set. Sending the entire set seemed pretty noisy

> ```
> BEGIN;
> SET z = 'a';
> ROLLBACK;
> ```
> 
> ```
> BEGIN;
> SET LOCAL z = 'a';
> COMMIT;
> ```
> 
> In both of these cases, the variable change message is sent, but after the txn ends the variable reverts to something else, and that change isn't sent.

I added test for these scenarios for the websocket protocol and they should work!

File: src/sql/src/session/vars.rs [hunk 3]
```diff
@@ -1389,10 +1395,22 @@ impl SessionVars {
 
     /// Commits or rolls back configuration parameter updates made via
     /// [`SessionVars::set`] since the last call to `end_transaction`.
```
Comment (mjibson) [inline src/sql/src/session/vars.rs:1397]: Add comment about what gets returned.

**Review by mjibson — APPROVED** [review]
Looks great. Love the new tests.

File: src/sql/src/session/vars.rs (hunk 3 above)
Comment (ParkMyCar) [reply src/sql/src/session/vars.rs:1398]: Done!

File: src/sql/src/session/vars.rs [hunk 4]
```diff
@@ -1274,6 +1274,12 @@ impl SessionVars {
             &STANDARD_CONFORMING_STRINGS,
             &TIMEZONE,
             &INTERVAL_STYLE,
+            // Including `cluster` and `cluster_replica` in the notify set is a Materialize
+            // extension. Doing so allows users to more easily identify where their queries
+            // will be executing, which is important to know when you consider the size of a
+            // cluster, what indexes are present, etc.
+            &*CLUSTER,
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/sql/src/session/vars.rs:1281]: Sounds good, I added it!

File: doc/user/content/integrations/websocket-api.md [hunk 5]
```diff
@@ -191,6 +192,30 @@ A single row result.
 Will only occur after a `Rows` message.
 The payload is an array of JSON values corresponding to the columns from the `Rows` message.
 
+#### `ParameterStatus`
+
+Executing a statement caused a session parameter to change. The parameters reported are:
+
+* `application_name`
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply doc/user/content/integrations/websocket-api.md]: Makes sense to me!

[the remark this answers is not in the document]
Comment (benesch) [reply doc/user/content/integrations/websocket-api.md]: I was going to say something similar. I think it makes more sense to add a new column like "protocol notified" or something like that to the session variable table in doc/user/layouts/shortcodes/session-variables.html. But if you're interested, making that change is probably best in its own PR, as I imagine the docs team will have some thoughts.

**Pull request merged**